module/zfs/zio.c

```diff
@@ -4102,6 +4132,10 @@ zio_dva_allocate(zio_t *zio)
 		ASSERT(zio->io_child_type > ZIO_CHILD_GANG);
 		zio->io_gang_leader = zio;
 	}
+	if (zio->io_flags & ZIO_FLAG_PREALLOCATED) {
+		ASSERT3U(zio->io_child_type, ==, ZIO_CHILD_GANG);
+		return (zio);
+	}
```

Comment:
I remember we discussed these ZIO changes, and same as before I am unhappy about adding more crutches there.  For the case when we have space allocated, etc, there is `zio_rewrite()` call.  If it misses something, it could be cleaner to add that instead.  Bypassing pipeline stages after getting into them IMHO a hack.

Reply:
I remember your issues, but as I said in the previous discussion, it's worse if you try to use rewrites. I tried it that way originally, and it was more invasive to make the necessary changes to the rewrite path than it was to just have the PREALLOCATED flag.

I have an idea for a slightly different way to implement the PREALLOCATED flag, though, let me try that and see what you think.